rust/lance-encoding/src/encodings/physical/buffers.rs

```diff
@@ -48,7 +64,230 @@ impl BufferEncoder for BitmapBufferEncoder {
         }
         let buffer = builder.finish().into_inner();
         let parts = vec![buffer];
-        let buffer = EncodedBuffer { parts };
+        let buffer = EncodedBuffer {
+            bits_per_value: 1,
+            parts,
+            bitpack_meta: None,
+        };
         Ok(buffer)
     }
 }
+
+#[derive(Debug, Default)]
+pub struct BitpackingBufferEncoder {}
+
+impl BufferEncoder for BitpackingBufferEncoder {
+    fn encode(&self, arrays: &[ArrayRef]) -> Result<EncodedBuffer> {
+        let mut num_bits = 0;
+        for arr in arrays {
+            let arr_max = compute_num_bits(arr.clone())?;
+            num_bits = num_bits.max(arr_max);
+        }
+
+        // TODO handle case where there are no arrays or all the arrays are full of zeros
+        if num_bits == 0 {
+            panic!("TODO handle zero length bitpacking")
+        }
+
+        let mut packed_arrays = vec![];
+        for arr in arrays {
+            let packed = pack_array(arr.clone(), num_bits)?;
+            packed_arrays.push(packed.into());
+        }
+
+        let data_type = arrays[0].data_type();
+        let bits_per_value = 8 * data_type.byte_width() as u64;
+
+        Ok(EncodedBuffer {
+            bits_per_value: num_bits,
+            parts: packed_arrays,
+            bitpack_meta: Some(pb::BitpackMeta {
+                uncompressed_bits_per_value: bits_per_value,
+            }),
+        })
```

Comment:
Do we want to conditionally bitpack based on the whether `num_bits` is less than "native num bits" if that makes sense?  E.g. if a number is using the full range then don't bitpack?

Reply:
Sure -- made this change